docs/design/backup_and_restore.md

```diff
@@ -0,0 +1,198 @@
+
+Changing the following fields will require running the modify command:
+
+* Backup URL parameters
+* Snapshot time
+
+#### Toggle Backup Paused
+
+In the `Toggle Backup Paused` stage, we will pause or resume the backup, if necessary. If the desired state is `paused` and the backup is not paused, this will run the `fdbbackup pause` command. If the desired state is `running` and the backup is paused, this will run the `fdbbackup resume` command.
+
+### Reconciling Restores
+
+The restore reconciler will run the following stages:
+
+* Toggle restore running
+* Check restore status
+
+We will use the `fdbrestore` tool to modify the restore state. We will use the restore status that is found within the database status to determine the current state of the backup.
+
+#### Toggle Restore Running
+
+In the `Toggle Restore Running` stage, we will start or stop the restore, if necessary. If the desired state is `stopped`, and there is a restore running on the configured tag, this will run the `fdbrestore abort` command. If the desired state is `running`, and there is no restore running on the configured tag, this will run the `fdbrestore start` command.
+
+#### Check Restore Status
+
+In the `Check Restore Status` stage, we will compare the running state and validate the following things:
+
+* The state of the restore matches the configured state in the spec
+* If the restore is configured to be running, it has completed
+
+If any of these checks fail, the restore will be marked as unreconciled.
+
+### Expiring and Deleting Backups
+
+In order to prevent backups from growing without bound, we have configuration options to expire backup data older than a certain age. This will be done through the `fdbbackup expire` command. Because this command blocks, and requires non-trivial resources to run, this will be run in a separate pod from the main controller. The operator will maintain an in-memory queue of backup expiration jobs, with one job for each backup. On start-up, it will populate the queue, and on reconciliation it will ensure that the backup has an entry in the expiration queue.  The operator will have a limit for the concurrency of backup expiration pods, with a default of 1. The operator will have a dedicated goroutine for monitoring this queue and creating pods. If the number of active backup expiration pods is less than the concurrency limit, the queueing goroutine will pop the first job off the queue, create a pod for running that backup expiration, and put a new entry on the end of the queue. If the number of active backup expiration pods is greater than or equal to the concurrency limit, this goroutine will sleep for 10 minutes and try again. It will also check to make sure that there is only one backup expiration running for a given backup at once. Pods that have started termination will not count against this concurrency limit.
```

Comment:
Do we want to make the sleep time configurable for starting new goroutines?

Reply:
Yes, it should be configurable.

Reply:
The concurrency limit is per Backup or global? I would interpret it like per Backup but I might be wrong.

Reply:
Since we are managing bare Pods (that's an assumption since we are talking about a `backup expiration pod limit`, please correct me) what happens when the expire is canceled (or actually the Pod is deleted) during that command? Would a `ReplicaSet` or a `Job` make more sense?

Reply:
If the pod is deleted, the operator will start a new expiry job for the next cluster in the queue.

Reply:
There are two different limits at play here:

* Global concurrency limit, which is configurable. This is the number of overall backup expiration jobs we can run at one time. Each job will target a single backup.
* Per-backup concurrency limit, which is hard-coded to 1 for safety reasons.

Reply:
We may need to think more about whether the "global" concurrency limit is scoped to the Kubernetes cluster or to the namespace.

Reply:
That's an interesting idea. Would we have one `BackupExpiry` object for each backup, or a single `BackupExpiryQueue` object? A single queue object sounds like it could be a problem since it would mean a resource of unbounded size, but if we are reconciling each expiry independently then we would have additional challenges around fairness. We could track the lastExpiryStartTime in the status on the BackupExpiry, and then on every reconciliation pull up the full list sorted by that time to determine if it is valid to start a new job for the given BackupExpiry. It could be an odd fit for the reconciliation model, though, because I don't think the BackupExpiry resource would ever reach a reconciled state. Maybe we could consider it "reconciled" if the last run completed successfully and the last expiry start time was in the last X hours, and rely on the controller periodically checking every resource to make sure jobs keep going.